### PR TITLE
Feature/columnstore2 column headers

### PR DESCRIPTION
**Pull request by gnusi**

### CODE REVIEW

File: core/index/merge_writer.cpp [hunk 1]
```diff
@@ -865,120 +967,146 @@ bool compute_field_meta(
...
-      payload = irs::get<irs::payload>(it);
-    }
+  auto write_column = [&column, &writer, this](auto& it) -> bool {
+    auto* payload = irs::get<irs::payload>(it);
 
-    while (it.next()) {
+    do {
       if (!progress_()) {
-        // stop was requsted
+        // Stop was requsted
```
Comment (Dronplane) [inline core/index/merge_writer.cpp:1045]:
```suggestion
        // Stop was requested
```

File: core/index/merge_writer.cpp [hunk 2]
```diff
@@ -865,120 +967,146 @@ bool compute_field_meta(
...
+  };
 
-  void reset(const column_info& info) {
-    if (!empty_ || info_ != info) {
-      info_ = info;
-      column_ = writer_->push_column(info);
-      empty_ = true;
+  do {
+    if (!write_column(*begin)) {
+      // Stop was requsted
```
Comment (Dronplane) [inline core/index/merge_writer.cpp:1061]:
```suggestion
      // Stop was requested
```

File: core/index/merge_writer.cpp [hunk 3]
```diff
@@ -865,120 +967,146 @@ bool compute_field_meta(
...
-  //   'false' otherwise
-  operator bool() const noexcept { return static_cast<bool>(writer_); }
+  if (it.next()) {
+    auto column = writer_->push_column(info, std::move(finalizer));
 
-  // returns 'true' if no data have been written to columnstore
-  bool empty() const noexcept { return empty_; }
+    do {
+      if (!progress_()) {
+        // Stop was requsted
```
Comment (Dronplane) [inline core/index/merge_writer.cpp:1093]:
```suggestion
        // Stop was requested
```

**Review by Dronplane — COMMENTED** [review]
So far so good. Only minor comments

File: core/formats/columnstore.cpp [hunk 4]
```diff
@@ -71,8 +72,39 @@
...
+      : name(std::move(rhs.name)), id(rhs.id) {
+    rhs.id = field_limits::invalid();
+  }
+
+  column_meta(const column_meta&) = delete ;
+  column_meta& operator=(const column_meta&) = delete;
+  column_meta& operator=(column_meta&& rhs) noexcept = delete;
+
+  bool operator==(const column_meta& rhs) const noexcept {
+    return name == rhs.name;
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/columnstore.cpp:96]: Yes, there might be columns with the same name but different ids in different segments.

File: core/formats/columnstore.cpp [hunk 5]
```diff
@@ -236,6 +268,210 @@ void read_compact(
...
+
+      out_ = index_output::make<encrypted_output>(
+        std::move(out_),
+        *out_cipher_,
+        blocks_in_buffer);
+    }
+  }
+}
+
+void meta_writer::write(string_ref name, field_id id) {
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/columnstore.cpp:339]: It's hard given `columnstore::writer::column` doesn't have any relationships with `column_meta`, i.e. it exposes `name()` and `id()` seperately.
TBH I'd treat the whole `columnstore` (including `column_meta`) as obsolete.

File: core/formats/columnstore.cpp [hunk 6]
```diff
@@ -236,6 +268,210 @@ void read_compact(
...
+    throw io_error(string_utils::to_string(
+      "failed to open file, path: %s",
+      filename.c_str()));
+  }
+
+  const auto checksum = format_utils::checksum(*in_);
+
+  constexpr const size_t FOOTER_LEN =
+      sizeof(uint64_t) // count
+    + sizeof(uint64_t) // max id
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/columnstore.cpp:412]: I thought about that, but relying on an alias is dangerous when dealing with on disk format, e.g. someone may decide to change it to `uint32_t` at some point which will break backwards compatibility as the consequence.

File: core/formats/columnstore.cpp [hunk 7]
```diff
@@ -236,6 +268,210 @@ void read_compact(
...
+    + format_utils::FOOTER_LEN;
+
+  // seek to start of meta footer (before count and max_id)
+  in_->seek(in_->length() - FOOTER_LEN);
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/columnstore.cpp:416]: Done

File: core/formats/columnstore.cpp [hunk 8]
```diff
@@ -2409,17 +2579,24 @@ bool reader::prepare(const directory& dir, const segment_meta& meta) {
...
+  columns.reserve(count);
+  for (field_id i = 0, size = static_cast<field_id>(count); i < size; ++i) {
     // read column properties
     const auto props = read_enum<ColumnProperty>(*stream);
     const auto factory_id = (props & (~CP_COLUMN_ENCRYPT));
 
     if (factory_id >= IRESEARCH_COUNTOF(COLUMN_FACTORIES)) {
       throw index_error(string_utils::to_string(
-        "Failed to load column id=" IR_SIZE_T_SPECIFIER ", got invalid properties=%d",
+        "Failed to load column id=" IR_SIZE_T_SPECIFIER "), got invalid properties=%d",
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/formats/columnstore.cpp:2599]: Fixed

File: core/index/merge_writer.cpp (hunk 1 above)
Comment (gnusi) [reply core/index/merge_writer.cpp:1045]: Fixed

File: core/index/merge_writer.cpp (hunk 2 above)
Comment (gnusi) [reply core/index/merge_writer.cpp:1061]: Fixed

File: core/index/merge_writer.cpp (hunk 3 above)
Comment (gnusi) [reply core/index/merge_writer.cpp:1093]: Fixed

**Pull request merged**